clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -0,0 +1,405 @@
+-- checks if the arguments are applied to a data constructor or primitive, and
+-- creates a value if it is fully applied. If the arguments are applied to a
+-- function, the first argument is applied.
+--
+evaluateAppWith
+  :: (Term -> State Env Value)
+  -> (PrimInfo -> [Either Value Type] -> State Env Value)
+  -> (Value -> Either Value Type -> State Env Value)
+  -> Term
+  -> Either Term Type
+  -> State Env Value
+evaluateAppWith eval evalPrim apply x y
+  | Data dc <- f
+  = if isFullyApplied (dcType dc) args
+       then VData dc <$> evalArgs args
+       else etaExpand term >>= eval
+
+  | Prim p <- f
+  , nArgs  <- length . fst $ splitFunForallTy (primType p)
+  = case compare (length args) nArgs of
+      LT -> etaExpand term >>= eval
+      EQ -> evalArgs args >>= evalPrim p
+      GT -> do
+        (pArgs, rArgs) <- splitAt nArgs <$> evalArgs args
+        primRes <- evalPrim p pArgs
+        foldM apply primRes rArgs
+
+  | otherwise
+  = do
+       evalF  <- eval f
+       evalAs <- evalArgs args
+       foldM apply evalF evalAs
+ where
+  term      = either (App x) (TyApp x) y
+  (f, args) = collectArgs term
+  evalArg   = bitraverse eval pure
+  evalArgs  = traverse evalArg
+
+-- | Default implementation for evaluating a letrec expression. This adds all
+-- bindings to the heap without eagerly evaluating them, then evaluates the
+-- body of the expression.
+--
+-- Bindings are not evaluated eagerly, as they may not be needed depending how
+-- the body of the expression evaluates (i.e. bindings may be in an elided
+-- alternative for a case expression).
+--
+evaluateLetrecWith
+  :: (Term -> State Env Value)
+  -> [LetBinding]
+  -> Term
+  -> State Env Value
+evaluateLetrecWith eval bs x =
+  State.withState addTerms (eval x)
+ where
+  addTerms env =
+    let terms = fmap (fmap Left) bs
+     in foldl' (flip $ uncurry insertLocal) env terms
+
+-- | Default implementation for evaluating a case expression. This elides the
+-- case expression if the branch taken can be determined statically, otherwise
+-- it evaluates all alternatives.
+--
+-- Only literals and data constructors can be matched on. While this is enough
+-- for simple frontends, more complex frontends may want more complex matching
+-- rules (e.g. matching on VData with literals or matching on primitives).
+--
+evaluateCaseWith
+  :: (Term -> State Env Value)
+  -> (Literal -> Pat -> Bool)
+  -> (DataCon -> Pat -> Bool)
+  -> Term
+  -> Type
+  -> [Alt]
+  -> State Env Value
+evaluateCaseWith eval matchLit matchData x ty ys = do
+  evalX <- eval x
+
+  -- TODO Strip ticks?
+  case evalX of
+    VLit l ->
+      case findMatchingAlt (matchLit l) ys of
+        Just (_, y) -> eval y
+        Nothing -> error ("evaluateCaseWith: No pattern for literal " <> show l)
+
+    VData dc _ ->
+      case findMatchingAlt (matchData dc) ys of
+        Just (_, y) -> eval y
+        Nothing -> error ("evaluateCaseWith: No pattern for data " <> show dc)
+
+    v -> do
```

Comment:
What about 
```haskell
case x of
  _ -> y
```
If `x` evaluates to a neutral, we can still elide the case, and only evaluate `y`. (If `x` evaluates to _|_, the case should evaluate to _|_ as well).

Reply:
Another thing to correct I've just noticed: identifiers / tyvars bound in `DataPat` need to be added to the environment. Currently those bindings are being lost and these things are becoming free vars